pkg/pulp_client/domains.go

```diff
@@ -0,0 +1,95 @@
+package pulp_client
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/content-services/content-sources-backend/pkg/config"
+	zest "github.com/content-services/zest/release/v2023"
+	"github.com/rs/zerolog/log"
+)
+
+const DefaultDomain = "default"
+
+func S3StorageConfiguration() map[string]interface{} {
+	loaded := config.Get().Clients.Pulp.CustomRepoObjects
+	s3Config := make(map[string]interface{})
+
+	s3Config["aws_default_acl"] = "@none None"
+	s3Config["aws_s3_region_name"] = loaded.Region
+	s3Config["endpoint_url"] = loaded.URL
+	s3Config["aws_access_key_id"] = loaded.AccessKey
+	s3Config["secret_key"] = loaded.SecretKey
+	s3Config["aws_storage_bucket_name"] = loaded.Name
+	return s3Config
+}
+
+func (r *pulpDaoImpl) LookupOrCreateDomain(name string) (*string, error) {
+	href, err := r.LookupDomain(name)
+	if err != nil {
+		return nil, err
+	}
+	if href != nil {
+		return href, nil
+	}
+	href, createErr := r.CreateDomain(name)
+	if err == nil {
+		return href, nil
+	} else {
+		// If we get an error, lookup the domain again to see if another request created it
+		//  if its still not there, return the create error
+		href, err := r.LookupDomain(name)
+		if href == nil || err != nil {
+			return nil, createErr
+		} else {
+			return href, nil
+		}
+	}
+}
+
+func (r *pulpDaoImpl) LookupDomain(name string) (*string, error) {
+	list, resp, err := r.client.DomainsAPI.DomainsList(r.ctx, "default").Name(name).Execute()
+	defer resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Results) == 0 {
+		return nil, nil
+	} else {
+		return list.Results[0].PulpHref, nil
+	}
+}
+
+// CreateRpmPublication Creates a Publication
+func (r *pulpDaoImpl) CreateDomain(name string) (*string, error) {
+	s3Storage := zest.STORAGECLASSENUM_STORAGES_BACKENDS_S3BOTO3_S3_BOTO3_STORAGE
+	localStorage := zest.STORAGECLASSENUM_PULPCORE_APP_MODELS_STORAGE_FILE_SYSTEM
+	var domain zest.Domain
+	if config.Get().Clients.Pulp.StorageType == config.STORAGE_TYPE_OBJECT {
+		config := S3StorageConfiguration()
+		domain = *zest.NewDomain(name, s3Storage, config)
+	} else {
+		emptyConfig := make(map[string]interface{})
+		emptyConfig["location"] = fmt.Sprintf("/var/lib/pulp/%v/", name)
+		domain = *zest.NewDomain(name, localStorage, emptyConfig)
+	}
+	domainResp, resp, err := r.client.DomainsAPI.DomainsCreate(r.ctx, DefaultDomain).Domain(domain).Execute()
+	if resp.Body != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		body := ""
+		if resp.Body != nil {
+			buf := new(bytes.Buffer)
+			_, err := buf.ReadFrom(resp.Body)
+			if err == nil {
+				body = buf.String()
+			} else {
+				log.Error().Err(err).Msg("Error reading body from failed domain creation.")
+			}
+		}
+		log.Info().Err(err).Str("body", body).Msg("Error creating domain")
```

Comment:
I think the error I was seeing was actually just this log, I notice it's actually just an Info log, so I'm assuming this is expected?

Reply:
Ahh, yes, so we have this bug:  https://github.com/pulp/pulpcore/issues/4183  so we can't treat 500s as fatal.  I figured it made sense then to just log all errors at info, since we don't know if its 'expected' or not.  I imagine once that issue is fixed, we could treat non-400s errors more severely and throw a proper 'error' log statement.

Reply:
I guess i could do it as Warn for now?  I'll file a task to followup on this too.

Reply:
i updated to warn